src/rdf.internal.ts

```diff
@@ -0,0 +1,585 @@
+    blankNodes: ReadonlyArray<Predicates | BlankNodeId>;
+  }>
+>;
+
+export type Predicates = Readonly<Record<IriString, Objects>>;
+
+export type Subject = Readonly<{
+  type: "Subject";
+  url: BlankNodeId | LocalNodeIri | IriString;
+  predicates: Predicates;
+}>;
+
+export type Graph = Readonly<
+  Record<BlankNodeId | LocalNodeIri | IriString, Subject>
+>;
+
+export type ImmutableDataset = Readonly<{
+  type: "Dataset";
+  graphs: Readonly<Record<IriString, Graph> & { default: Graph }>;
+}>;
+
+/**
+ * Runtime freezing might be too much overhead;
+ * if so, this function allows us to replace it by a function
+ * that merely marks its input as Readonly<> for static analysis.
+ */
+export const freeze: typeof Object.freeze = Object.freeze;
+
+export function fromRdfJsDataset(
+  rdfJsDataset: RdfJs.DatasetCore
+): ImmutableDataset {
+  const dataset: ImmutableDataset = {
+    graphs: { default: {} },
+    type: "Dataset",
+  };
+
+  const quads = Array.from(rdfJsDataset);
+
+  const chainBlankNodes = getChainBlankNodes(quads);
+
+  // Quads with chain Blank Nodes as their Subject will be parsed when those
+  // Blank Nodes are referred to in an Object. See `addRdfJsQuadToObjects`.
+  const quadsWithoutChainBlankNodeSubjects = quads.filter((quad) =>
+    chainBlankNodes.every(
+      (chainBlankNode) => !chainBlankNode.equals(quad.subject)
+    )
+  );
+
+  return quadsWithoutChainBlankNodeSubjects.reduce(
+    (datasetAcc, quad) =>
+      addRdfJsQuadToDataset(datasetAcc, quad, {
+        otherQuads: quads,
+        chainBlankNodes: chainBlankNodes,
+      }),
+    dataset
+  );
+}
+
+export function getChainBlankNodes(quads: RdfJs.Quad[]): RdfJs.BlankNode[] {
+  const blankNodeSubjects = quads
+    .map((quad) => quad.subject)
+    .filter(isBlankNode);
+  const blankNodeObjects = quads.map((quad) => quad.object).filter(isBlankNode);
+  function everyNodeTheSame(nodes: RdfJs.Term[]): boolean {
+    // This could potentially be made more performant by mapping every term
+    // to their value and using native JS comparisons, assuming every node is
+    // either a Blank or a Named Node.
+    return nodes.every((otherNode) =>
+      nodes.every((anotherNode) => otherNode.equals(anotherNode))
+    );
+  }
+
+  const cycleBlankNodes: RdfJs.BlankNode[] = [];
+  blankNodeObjects.forEach((blankNodeObject) => {
+    cycleBlankNodes.push(...getCycleBlankNodes(blankNodeObject, [], quads));
+  });
+
+  // Get Blank Nodes that are used to provide nested values for a single Subject,
+  // which we'll represent as nested values as well
+  // (this allows us to avoid generating a non-deterministic, ad-hoc identifier
+  // for those Blank Nodes):
+  const chainBlankNodes = blankNodeSubjects
+    .concat(blankNodeObjects)
+    .filter((blankNode) => {
+      if (
+        cycleBlankNodes.some((cycleBlankNode) =>
+          cycleBlankNode.equals(blankNode)
+        )
+      ) {
+        return false;
+      }
```

Comment:
Given the complexity I definitely should have done this the first time around. Of course, since it's recursion it's still fairly complex, but I gave it a go here: cb213c7b.